devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp

```diff
@@ -303,75 +303,258 @@ bool WholeBodyDynamicsDevice::openEstimator(os::Searchable& config)
+                return false;
+            }
         }
-    }
 
+        // We build the overrideContactFramesIdx vector
+        std::vector<iDynTree::FrameIndex> overrideContactFramesIdx;
+        overrideContactFramesIdx.clear();
+        for(size_t i=0; i < overrideContactFrames.size(); i++)
+        {
+            iDynTree::FrameIndex idx = estimator.model().getFrameIndex(overrideContactFrames[i]);
 
-    // Let's check that every submodel has a default contact position
-    for(size_t subModelIdx = 0; subModelIdx < nrOfSubModels; subModelIdx++)
-    {
-        if( subModelIndex2DefaultContact[subModelIdx] == iDynTree::FRAME_INVALID_INDEX )
+            if( idx == iDynTree::FRAME_INVALID_INDEX )
+            {
+                yWarning() << "Frame " << overrideContactFrames[i] << " not found in the model, discarding it";
+            }
+            else
+            {
+                overrideContactFramesIdx.push_back(idx);
+            }
+        }
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// The frames from a single submodel should be no more than 1 if the type is FULL_WRENCH, no more than 3 if the type is PURE_FORCE, and no more than 6 if the type is PURE_FORCE_WITH_KNOWN_DIRECTION.
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+        // For each submodel, we find the first suitable contact frame
+        // This is a n^2 algorithm, but given that is just used in
+        // configuration it should be ok
+        size_t nrOfSubModels = estimator.submodels().getNrOfSubModels();
+
+        //// We indicate with FRAME_INVALID_INDEX the fact that we still don't have a override contact for the given submodel
+        subModelIndex2OverrideContact.resize(overrideContactFramesIdx.size());
+
+        subModelVarSize.resize(nrOfSubModels);
+
+        for(size_t i=0; i < overrideContactFramesIdx.size(); i++)
         {
-            yError() << "wholeBodyDynamics : openDefaultContactFrames : missing default contact for submodel composed by the links: ";
-            const iDynTree::Traversal & subModelTraversal = estimator.submodels().getTraversal(subModelIdx);
-            for(iDynTree::TraversalIndex i = 0; i < (iDynTree::TraversalIndex) subModelTraversal.getNrOfVisitedLinks(); i++)
+            // We indicate with FRAME_INVALID_INDEX the fact that we still don't have a override contact for the given submodel
+            subModelIndex2OverrideContact[i].resize(nrOfSubModels,iDynTree::FRAME_INVALID_INDEX);
+
+            size_t subModelIdx = estimator.submodels().getSubModelOfFrame(estimator.model(),overrideContactFramesIdx[i]);
+
+            //check if adding the current contact to this subModel will increase the no of variable to more than 6
+            if(subModelVarSize[subModelIdx]+wrench[i].nrUnknownsInExtWrench > 6)
```

Comment:
@fjandrad 

> You are enforcing that the number of unkowns remains to be 6, so you can find a unique solution?

Yes, following the recommendation of the [theory paper](https://www.researchgate.net/publication/236152161_Contact_Force_Estimations_Using_Tactile_Sensors_and_Force_Torque_Sensors), see section 3 Discussion and conclusion.




> I think that by default there was a behavior to evenly distribute the forces in case of more than one fully unkown contact. Not ideal, but it allowed having more contacts with some heuristic

I am not aware of that, do you mean when there are 2 full-wrench contacts? and the unknowns count will rise to 12?

The theory paper was talking about evenly distributing the forces. But in the case of 2 pure-force contacts in a submodel. In which case, the number of variables is 6 but the rank of the A matrix drops to 5 (related [comment](https://github.com/dic-iit/element_ironcub-estimation/issues/36#issuecomment-591534146)). So they performed statistical analysis in this particular case if I understood well.

> With your change that behavior would be changed. If we agree this is the new behavior it would make change to add it in some changelog or somewhere so that this is made known to the users.

What is currently implemented before the PR is that (in case of taking the contacts information from the `defaultContactFrame` list), the device assigns only one contact force for each submodel and assumes it's full-wrench. See this part:
https://github.com/robotology/whole-body-estimators/blob/d63c82e8c65d483e218e5c42ecd834bb54612ffd/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp#L341
And documentation:
https://github.com/robotology/whole-body-estimators/blob/d63c82e8c65d483e218e5c42ecd834bb54612ffd/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.h#L131

Reply:
Yes, but from the code point of view I think there is nothing preventing of putting two contacts in the same submodel using the list of default contacts.

Reply:
Nvm I saw it checks if there is a valid contact already so yes there would only be one contact in case of not using the skin.

Reply:
> Nvm I saw it checks if there is a valid contact already so yes there would only be one contact in case of not using the skin.

Yes, this is done by https://github.com/robotology/whole-body-estimators/blob/c0ca35e6d7b86a99ed6043aec69462c1cd10f066/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp#L372